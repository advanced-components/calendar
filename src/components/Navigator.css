.mini-month-navigator {
  margin-top: 15px;
  width: 100%;
}

.mini-month-navigator tr {
  height: 30px;
}

.mini-month-navigator td.today {
  color: white;
  background: #333;
  border-radius: 50%;
}

.mini-month-navigator td {
  width: 36px;
  height: 36px;
  text-align: center;
}

.mini-month-navigator .day {
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  border-radius: 50%;
  cursor: pointer;
}

.mini-month-navigator .day.offset {
  color: #eee;
}

.mini-month-navigator .day:hover {
  background: #f1f3f5;
}

.mini-month-navigator .current {
  background: #868e96;
  color: white;
}

.mini-month-navigator .current:hover {
  background: #495057;
  color: white;
}

.mini-month-navigator .today {
  background: #495057;
  color: white;
}

.mini-month-navigator .today:hover {
  background: #212529;
}
