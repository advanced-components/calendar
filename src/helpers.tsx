export const getMonthMatrix = (date: Date) => {
  const calendarMatrix: Date[][] = [];
  const month = date.getMonth();
  const startDate = new Date(2019, month, 1);
  const endDate = new Date(2019, month + 1, 0);

  //@TODO !!!!!
  if (startDate.getDay() != 1) {
    startDate.setDate(startDate.getDay() - 8);
  }
  if (endDate.getDay() != 1) {
    endDate.setDate(endDate.getDate() + (7 - endDate.getDay()));
  }
  endDate.setDate(endDate.getDate() + 7);
  //@TODO !!!!!

  let week = [];
  while (startDate <= endDate) {
    week.push(new Date(startDate));
    if (week.length === 7) {
      calendarMatrix.push(week);
      week = [];
    }
    startDate.setDate(startDate.getDate() + 1);
  }
  return calendarMatrix;
};

export const getDaysOfYearArray = () => {
  var daysOfYear: Date[] = [];
  for (
    var d = new Date(2019, 0, 1);
    d <= new Date(2019, 11, 31);
    d.setDate(d.getDate() + 1)
  ) {
    daysOfYear.push(new Date(d));
  }
  while (daysOfYear[0].getDay() != 1) {
    const date = new Date(daysOfYear[0]);
    const previousDate = date.setDate(date.getDate() - 1);
    daysOfYear.unshift(new Date(previousDate));
  }
  while (daysOfYear[daysOfYear.length - 1].getDay() != 0) {
    const date = new Date(daysOfYear[0]);
    const nextDate = date.setDate(date.getDate() + 1);
    daysOfYear.push(new Date(nextDate));
  }
  return daysOfYear;
};

// export const getDaysArray = (year?: Date) => {
//   const daysArray: Date[] = [];
//   for (let month = 0; month < MONTHS.length; month++) {
//     for (let day = 0; day < MONTHS[month]; day++) {
//       const date = new Date(2019, month, day);
//       daysArray.push(date);
//     }
//   }
//   while (daysArray[0].getDay() != 1) {
//     const date = daysArray[0];
//     const previousDate = new Date(date.setDate(date.getDate() - 1));
//     daysArray.unshift(previousDate);
//   }
//   while (daysArray[daysArray.length - 1].getDay() != 0) {
//     const date = daysArray[0];
//     const nextDate = new Date(date.setDate(date.getDate() + 1));
//     daysArray.push(nextDate);
//   }
//   return daysArray;
// };

function getMonthArray(date: Date) {
  const month: any = [[]];
  const firstDay = new Date(date.getFullYear(), date.getMonth(), 1);
  const lastDay = new Date(date.getFullYear(), date.getMonth() + 1, 0);
  const start = firstDay.getDay();
  const end = lastDay.getDate();

  for (let i = 1; i < start; i++) {
    month[0].push('');
  }

  for (let i = 1; i <= end; i++) {
    const currentWeek = month[month.length - 1];
    const day = String(i);
    if (currentWeek.length === 7) {
      month.push([day]);
    } else {
      currentWeek.push(day);
    }
  }

  return month;
}

function getDateTitle(date: Date) {
  const options = { year: 'numeric', month: 'long' };
  return date.toLocaleDateString('en-EN', options);
}

function getMondayDate(date: Date) {
  const day = date.getDay();
  const diff = date.getDate() - day + (day === 0 ? -6 : 1);
  return new Date(date.setDate(diff)).getDate();
}

function formatDate(date: number) {
  const d = new Date(date);
  let month = `${d.getMonth() + 1}`;
  let day = `${d.getDate()}`;
  const year = d.getFullYear();

  if (month.length < 2) month = `0${month}`;
  if (day.length < 2) day = `0${day}`;

  return [year, month, day].join('-');
}

export { getMonthArray, getMondayDate, getDateTitle, formatDate };
