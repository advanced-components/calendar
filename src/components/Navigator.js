import React from 'react';
import './Navigator.css';

export const Navigator = () => (
  <div className="navigation">
    <button type="button">TODAY</button>
    <button type="button">{'<'}</button>
    <button type="button">{'>'}</button>
  </div>
);

export default Navigator;
