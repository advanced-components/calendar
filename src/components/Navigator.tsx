import React from 'react';
import { getMonthMatrix, getDateTitle } from '../helpers';

import './Navigator.css';

export const Navigator: React.FunctionComponent<{
  date: Date;
  onDateChange: Function;
}> = ({ date, onDateChange }) => {
  const monthMatrix = getMonthMatrix(date);
  const month_table = monthMatrix.map((week: any, index: number) => (
    <tr key={index}>
      {week.map((monthMatrixDate: Date, index: number) => {
        const today = new Date();
        //@TODO !!!!!
        let className = 'day';
        if (date.getMonth() != monthMatrixDate.getMonth())
          className += ' offset ';
        if (
          monthMatrixDate.getMonth() === date.getMonth() &&
          monthMatrixDate.getDate() === date.getDate()
        )
          className += ' current ';
        if (
          monthMatrixDate.getMonth() === today.getMonth() &&
          monthMatrixDate.getDate() === today.getDate()
        )
          className += ' today ';
        //@TODO !!!!!
        return (
          <td key={index}>
            <span
              className={className}
              onClick={() => onDateChange(monthMatrixDate)}
            >
              {monthMatrixDate.getDate()}
            </span>
          </td>
        );
      })}
    </tr>
  ));

  return (
    <div className='sidebar-month' id='sidebar-month'>
      <h2 className='date-info'>{getDateTitle(date)}</h2>
      <table className='mini-month-navigator'>
        <tbody>{month_table}</tbody>
      </table>
    </div>
  );
};
