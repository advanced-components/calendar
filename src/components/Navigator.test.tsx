import React from 'react';
import renderer from 'react-test-renderer';
import { Navigator } from './Navigator';

import { getMonthArray, getDateTitle } from '../helpers';

const date = new Date(0);
const defaultProps = {
  monthName: getDateTitle(date),
  monthArray: getMonthArray(date),
  currentDay: 1
};

it('renders correctly', () => {
  const component = renderer.create(<Navigator {...defaultProps} />).toJSON();
  expect(component).toMatchSnapshot();
});
