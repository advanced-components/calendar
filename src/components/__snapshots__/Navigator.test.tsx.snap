// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Navigator /> renders correctly 1`] = `
<div
  className="sidebar-month"
  id="sidebar-month"
>
  <h2
    className="date-info"
  >
    January 1970
  </h2>
  <table
    className="mini-month-navigator"
  >
    <tbody>
      <tr>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            27
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            28
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            29
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            30
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            31
          </span>
        </td>
        <td>
          <span
            className="day current "
            onClick={[Function]}
          >
            1
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            2
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            3
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            4
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            5
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            6
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            7
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            8
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            9
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            10
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            11
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            12
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            13
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            14
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            15
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            16
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            17
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            18
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            19
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            20
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            21
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            22
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            23
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            24
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            25
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            26
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            27
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            28
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            29
          </span>
        </td>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            30
          </span>
        </td>
      </tr>
      <tr>
        <td>
          <span
            className="day"
            onClick={[Function]}
          >
            31
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            1
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            2
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            3
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            4
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            5
          </span>
        </td>
        <td>
          <span
            className="day offset "
            onClick={[Function]}
          >
            6
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
